lib/ProductOpener/Producers.pm

```diff
@@ -0,0 +1,564 @@
+						push @{$columns_fields_ref->{$column}{examples}}, $value;
+					}
+				}
+
+				# content
+				if ($value =~ /[0-9]/) {
+					$columns_fields_ref->{$column}{numbers}++;
+				}
+				if ($value =~ /[a-z]/i) {
+					$columns_fields_ref->{$column}{letters}++;
+				}
+				if ($value =~ /[0-9].*[a-z]/i) {
+					$columns_fields_ref->{$column}{both}++;
+				}
+			}
+
+			$col++;
+		}
+
+		$row++;
+	}
+
+	# Load previously assigned fields by the user_agent
+
+	my $all_columns_fields_ref = {};
+
+	if (defined $owner) {
+		$all_columns_fields_ref = retrieve("$data_root/import_files/$owner/all_columns_fields.sto");
+	}
+
+	# Match known column names to OFF fields
+
+	foreach my $column (@$headers_ref) {
+
+		my ($field, $value_unit, $tag);
+
+		if (defined $all_columns_fields_ref->{$field}) {
+
+			$field = $all_columns_fields_ref->{$column}{field};
+			$value_unit = $all_columns_fields_ref->{$column}{value_unit};
+			$tag = $all_columns_fields_ref->{$column}{tag};
+		}
+		else {
+
+			my $column_id = get_string_id_for_lang("no_language", $column);
+
+			if ($column_id =~ /^(code|barcode|ean|ean13|ean-13)$/) {
+				$field = "code";
+			}
+
+			# If we don't know if the column contains value + unit, value, or unit,
+			# try to guess from the content of the column
+			if (not defined $value_unit) {
+				if ($columns_fields_ref->{$column}{both}) {
+					$value_unit = "value_unit";
+				}
+				elsif ($columns_fields_ref->{$column}{numbers}) {
+					$value_unit = "value";
+				}
+				elsif ($columns_fields_ref->{$column}{letters}) {
+					$value_unit = "unit";
+				}
+			}
+		}
+
+		$columns_fields_ref->{$column}{field} = $field;
+		$columns_fields_ref->{$column}{value_unit} = $value_unit;
+		$columns_fields_ref->{$column}{tag} = $tag;
+
+		delete $columns_fields_ref->{$column}{existing_examples};
+	}
+
+	return $columns_fields_ref;
+}
+
+
+# Generate an array of options for select2
+
+sub generate_import_export_columns_groups_for_select2($$) {
+
+	my $fields_groups_ref = shift;
+	my $lcs_ref = shift; # array of language codes
+
+	# Sample select2 groups and fields definition format from Config.pm:
+	my $sample_fields_groups_ref = [
+		["identification", ["code", "producer_product_id", "producer_version_id", "lang", "product_name", "generic_name",
+			"quantity_value_unit", "net_weight_value_unit", "drained_weight_value_unit", "volume_value_unit", "packaging",
+			"brands", "categories", "categories_specific", "labels", "labels_specific", "countries", "stores"]
+		],
+		["origins", ["origins", "origin", "manufacturing_places", "producer", "emb_codes"]
+		],
+		["ingredients", ["ingredients_text", "allergens", "traces"]
+		],
+		["nutrition"],
+		["nutrition_other"],
+		["other", ["conservation_conditions", "warning", "preparation", "recipe_idea", "recycling_instructions_to_recycle", "recycling_instructions_to_discard", "customer_service", "link"]
+		],
+	];
+
+	# Create an options array for select2
+
+	my $select2_options_sample = <<JSON
```

Comment:
Why don't you just build this with `JSON:PP`/`JSON:XS`? I'd avoid building JSON manually, if possible.

Reply:
This is just a sample JSON expected by select2, it's not actually used anywhere, it's more like a comment, to explain what kind of JSON structure is expected by select2. I can make it an actual comment instead.